specs/spec.md

```diff
@@ -0,0 +1,319 @@
+
+Every hobbit message contains the following fields: 
+
+| Field | Definition | Validity |
+|:------:|----------|:----:|
+| `version` | Defines the EWP version number e.g. `0.2`. | `(\d+\.)(\d+)` |
+| `protocol` | Defines the [protocol](#protocols). | `(RPC\|GOSSIP\|PING)` |
+| `header` | Defines the header | payload |
+| `body` | Defines the body | payload |
+
+## Transport
+
+<!-- @TODO DISCUSS TCP AND SHIT -->
+
+## Protocols
+
+Hobbits defines 3 protocols that dictate how messages are interpreted and responded to.
+
+### RPC
+
+<!-- @TODO WRITE SOME SHIT -->
+
+#### Envelope
+
+```python
+{ 
+  'method_id': 'uint16' ## byte representing the method
+  'id': 'uint64' ## id of the request
+  'body': 'bytes' ## body of the request itself
+}
+```
+
+The `body` field contains an [SSZ](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md) encoded byte array. 
+
+##### Example
+
+Below you will find an example `RPC` request.
+
+```
+EWP 0.2 RPC 0 65
+{
+  "method_id": 0x01,
+  "id": 1,
+  "body": {
+    "bodies": "0x2a00000000000000"
+  }
+}
+```
+#### RPC Methods
+
+There are two types of RPC methods: one that requests data and one that responds with data.
+
+* Peers may request blocks and headers from other peers, possibly in response to a `GOSSIP` message. 
+* Peers may request blocks repeatedly from the same peers.
+* Other peers may respond on a best effort basis with header and block data.
+* There is no SLA for responding. 
+
+##### Handshake
+
+###### `0x00` HELLO
+
+Upon discovering each other, nodes may exchange `HELLO` messages.
+
+Nodes may send `HELLO` to other peers when they exchange messages for the first time or when their state changes to let them know new blocks are available.
+
+Nodes can send `HELLO` messages to each other to exchange information on their status:
+
+```python
+{
+  'network_id': 'uint8' ## the ID of the network (1 for mainnet, and some predefined number for a testnet)
+  'chain_id': 'uint8' ## the ID of the chain (1 for ETH)
+  'latest_finalized_root': 'bytes32' ## the hash of the latest finalized root
+  'latest_finalized_epoch': 'uint64' ## the number of the latest finalized epoch
+  'best_root': 'bytes32' ## the hash of the best root this node can offer
+  'best_slot': 'uint64' ## the number of the best slot this node can offer
+}
+```
+
+Upon receiving a `HELLO` message, the node should reply with a `HELLO` message.
+
+###### `0x01` GOODBYE
+
+Nodes may signal to other nodes that they are going away by sending a `GOODBYE` message:
+
+```python
+{
+  'reason': 'uint64' ## an optional reason code up to the client
+}
+```
+
+The reason given is optional. Reason codes are up to each client and should not be trusted.
+
+Upon receiving a `GOODBYE` message, no response is necessary.
+
+##### `0x02` GET_STATUS
+
+Nodes may exchange metadata information using a `GET_STATUS` message.
+<!--is this necessary?-->
+A `GET_STATUS` request may be sent in response to receiving a `GOSSIP` message:
```

Comment:
This actually might be a good place to return (fork choice) head block.

Reply:
@jrhea

Reply:
@cemozerr open a new issue and add more fields to status. You can throw a bunch of metadata in there.